libs/checkpoint-postgres/langgraph/store/postgres/base.py

```diff
@@ -145,60 +325,162 @@ def _get_batch_PUT_queries(
+
             values_str = ",".join(values)
             query = f"""
                 INSERT INTO store (prefix, key, value, created_at, updated_at)
                 VALUES {values_str}
                 ON CONFLICT (prefix, key) DO UPDATE
-                SET value = EXCLUDED.value, updated_at = CURRENT_TIMESTAMP
+                SET value = EXCLUDED.value,
+                    updated_at = CURRENT_TIMESTAMP
             """
             queries.append((query, insertion_params))
 
-        return queries
+            if vector_values:
+                values_str = ",".join(vector_values)
+                query = f"""
+                    INSERT INTO store_vectors (prefix, key, field_name, embedding, created_at, updated_at)
+                    VALUES {values_str}
+                    ON CONFLICT (prefix, key, field_name) DO UPDATE
+                    SET embedding = EXCLUDED.embedding,
+                        updated_at = CURRENT_TIMESTAMP
+                """
+                embedding_request = (query, embedding_request_params)
 
-    def _get_batch_search_queries(
+        return queries, embedding_request
+
+    def _prepare_batch_search_queries(
         self,
         search_ops: Sequence[tuple[int, SearchOp]],
-    ) -> list[tuple[str, Sequence]]:
-        queries: list[tuple[str, Sequence]] = []
-        for _, op in search_ops:
-            query = """
+    ) -> tuple[
+        list[tuple[str, list[Union[None, str, list[float]]]]],  # queries, params
+        list[tuple[int, str]],  # idx, query_text pairs to embed
+    ]:
+        queries = []
+        embedding_requests = []
+
+        for idx, (_, op) in enumerate(search_ops):
+            base_query = """
                 SELECT prefix, key, value, created_at, updated_at
                 FROM store
                 WHERE prefix LIKE %s
             """
             params: list = [f"{_namespace_to_text(op.namespace_prefix)}%"]
+            needs_vector_search = False
+
+            if op.query and self.embedding_config:
+                needs_vector_search = True
+                embedding_requests.append((idx, op.query))
+
+                score_expr = _get_distance_operator(self)
+                vector_type = (
+                    cast(PostgresEmbeddingConfig, self.embedding_config)
+                    .get("index_config", self._get_default_index_config())
+                    .get("vector_type", "vector")
+                )
+
+                if (
+                    vector_type == "bit"
+                    and self.embedding_config.get("distance_type") == "hamming"
+                ):
+                    score_expr = score_expr % ("%s", self.embedding_config["dims"])
+                else:
+                    score_expr = score_expr % ("%s", vector_type)
+
+                vectors_per_doc_estimate = self.embedding_config[
+                    "__estimated_num_vectors"
+                ]
+                expanded_limit = (op.limit * vectors_per_doc_estimate * 2) + 1
```

Comment:
flagging this query as being approximate.

Ideally we:
1. Get all values matching filters (in namespace prefix, etc.)
2. Get all the store_vectors for those docs
3. Sort by score
4. Select the top K items based on max(score) over the vectors

If we materialize a large view  in (1), that makes the search index less useful (i think) and in general could use a ton of memory.

As a compromise right now I'm running this in. CTE and returning up to K * N where N is twice the estimated number of vectors per doc.